src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/json/EventTransformer.scala

```diff
@@ -67,23 +67,95 @@ object EventTransformer {
-        if (latitude.nonEmpty && longitude.nonEmpty) {
-          JObject("geo_location" -> JString(s"$latitude,$longitude"))
-        } else {
-          JObject()
-        }
-      }
+      val geoLocation: JObject = getGeoLocationField(event)
+      if (flatten) convertEvent(event.toList, geoLocation)
+      else getJObjectWithNestedStructures(event)
+    }
+  }
+
+  private def getGeoLocationField(event: Array[String]): JObject = {
+    val latitude = event(GeopointIndexes.latitude)
+    val longitude = event(GeopointIndexes.longitude)
+    if (latitude.nonEmpty && longitude.nonEmpty) {
+      JObject("geo_location" -> JString(s"$latitude,$longitude"))
+    } else {
+      JObject()
+    }
+  }
+
+  private def getJObjectWithNestedStructures(event: Array[String]): Validated[(Set[InventoryItem], JObject)] = {
+    for {
+      jObject <- convertTsvEventToNestedJObject(event)
+    } yield (Set.empty[InventoryItem], getGeoLocationField(event) ~ jObject)
+  }
 
-      convertEvent(event.toList, geoLocation)
+  /**
+    * Event has been checked for size
+    *
+    * @param event the event array
+    * @return either a list of errors or the JObject representation
+    */
+  private def convertTsvEventToNestedJObject(event: Array[String]): Validated[JObject]= {
+    val validJObjects = FieldNames.zip(event).zip(FieldTypes).map {
+      case ((fieldName: String, fieldValue: String), fieldType: TsvToJsonConverter) => fieldParser(fieldName, fieldValue, fieldType)
+    }
+    validJObjects.partition(_.isLeft) match {
+      case (errors, _) if (errors.nonEmpty) => Left(for (Left(error) <- errors) yield error)
+      case (_, kvs) => Right(JObject(for (Right(kv) <- kvs) yield kv))
     }
   }
 
+  private def fieldParser(fieldName: ShredFieldName, fieldValue: String, fieldType: TsvToJsonConverter): Either[String, (String, JValue)] =
+    try {
+      Right(fieldName -> stringToJValue(fieldType, fieldValue))
+    } catch {
+      case nfe: NumberFormatException => Left(nfe.toString)
+      case bfe: BooleanFormatException => Left(bfe.toString)
+    }
+
+  private def stringToJValue(fieldType: TsvToJsonConverter, fieldValue: String): JValue =
+    if (isJObjectType(fieldType)) stringToJObjectConverter(fieldType, fieldValue)
+    else stringToScalarJValueConverter(fieldType, fieldValue)
+
+  private def isJObjectType(fieldType: TsvToJsonConverter): Boolean =
+    fieldType == CustomContextsField || fieldType == DerivedContextsField || fieldType == UnstructField
+
+  private def stringToJObjectConverter(fieldType: TsvToJsonConverter, fieldValue: String): JValue = {
+    if(fieldValue.isEmpty) JNothing
+    else if(isJObjectType(fieldType)) parse(fieldValue)
+    else throw new IllegalArgumentException(s"Unknown type: $fieldType")
+  }
+
+  private def stringToScalarJValueConverter(fieldType: TsvToJsonConverter, fieldValue: String): JValue = {
+    if (fieldValue.isEmpty) JNull
+    else if (fieldType == StringField) JString(fieldValue)
```

Comment:
Cannot we switch to pattern-match here? It will make me more confident regarding non-exhaustive match.

Reply:
We could, but the real gain would be (in a theoretical future refactoring) to convert StringType et c. to types.

Reply:
Makes sense.